Userland/Libraries/LibJS/Runtime/ECMAScriptFunctionObject.cpp

```diff
@@ -785,50 +789,68 @@ void async_block_start(VM& vm, T const& async_body, PromiseCapability const& pro
 
     // 3. Set the code evaluation state of asyncContext such that when evaluation is resumed for that execution context the following steps will be performed:
     auto execution_steps = NativeFunction::create(realm, "", [&realm, &async_body, &promise_capability, &async_context](auto& vm) -> ThrowCompletionOr<Value> {
-        // a. Let result be the result of evaluating asyncBody.
         Completion result;
-        if (auto* bytecode_interpreter = vm.bytecode_interpreter_if_exists()) {
-            // FIXME: Cache this executable somewhere.
-            auto maybe_executable = Bytecode::compile(vm, async_body, FunctionKind::Async, "AsyncBlockStart"sv);
-            if (maybe_executable.is_error())
-                result = maybe_executable.release_error();
-            else
-                result = bytecode_interpreter->run_and_return_frame(realm, *maybe_executable.value(), nullptr).value;
-        } else {
-            result = async_body->execute(vm.interpreter());
+
+        // a. If asyncBody is a Parse Node, then
+        if constexpr (!IsCallableWithArguments<T, Completion>) {
```

Comment:
can we use `IsSame<T, U>` here instead?

Reply:
Do you mean `IsSame<ASTNode>` type thing? Or `IsSame<SafeFunction<Completion()>`? I spent quite a while trying to figure out the best way of doing this :laughing: Haven't worked with templates too much

Reply:
I was thinking something like `IsSame<T, NonnullRefPtr<Statement>>`, but probably doesn't work with templates. Nevermind!

Reply:
Yeah, see the discussion in `#development`, I did try to do it that way but Ali ended up suggesting I just do this instead